pyannote/database/custom.py

```diff
@@ -230,113 +148,234 @@ def resolve_path(path: Text, database_yml: Path) -> Path:
     raise FileNotFoundError(msg)
 
 
-def add_custom_protocols():
-    """Update pyannote.database.{DATABASES|TASKS} with custom & meta protocols
+def meta_subset_iter(
+    meta_database: Text,
+    meta_task: Text,
+    meta_protocol: Text,
+    meta_subset: Text,
+    subset_entries: Dict,
+    database_yml: Path,
+):
+    """Meta-protocol method that iterates over a subset
 
-    Returns
-    -------
-    pyannote.database.DATABASES
-    pyannote.database.TASKS
+    Parameters
+    ----------
+    meta_database : str
+        "X"
+    meta_task : str
+        Task name (e.g. SpeakerDiarization, SpeakerVerification)
+    meta_protocol : str
+        Protocol name (e.g. MyProtocol)
+    meta_subset : {"train", "development", "test"}
+        Subset
+    subset_entries : dict
+        Subset entries.
+            Etape.SpeakerDiarization.TV: [train]
+            REPERE.SpeakerDiarization.Phase1: [train, development]
+            REPERE.SpeakerDiarization.Phase2: [train, development]
     """
 
-    from .config import get_database_yml
+    # this is imported here to avoid circular imports
+    from . import get_protocol
 
-    try:
-        database_yml = get_database_yml()
-        with open(database_yml, 'r') as fp:
-            config = yaml.load(fp, Loader=yaml.SafeLoader)
+    for protocol, subsets in subset_entries.items():
+        partial_protocol = get_protocol(protocol)
 
-    except FileNotFoundError:
-        config = dict()
+        for subset in subsets:
+            # TODO. get rid of this ugly mapping in favor of "train" -> "train_iter"
+            subset_short = {"train": "trn", "development": "dev", "test": "tst"}[subset]
+            method_name = f"{subset_short}_iter"
+            for file in getattr(partial_protocol, method_name)():
+                yield file
+
+
+def subset_iter(
+    database: Text,
+    task: Text,
+    protocol: Text,
+    subset: Text,
+    entries: Dict,
+    database_yml: Path,
+):
+    """
 
-    databases = config.get('Protocols', dict())
+    Parameters
+    ----------
+    database : str
+        Database name (e.g. MyDatabase)
+    task : str
+        Task name (e.g. SpeakerDiarization, SpeakerVerification)
+    protocol : str
+        Protocol name (e.g. MyProtocol)
+    subset : {"train", "development", "test"}
+        Subset
+    entries : dict
+        Subset entries.
 
-    # make sure meta-protocols are processed last (relies on the fact that
-    # dicts are iterated in insertion order since Python 3.6)
-    x = databases.pop('X', None)
-    if x is not None:
-        databases['X'] = x
 
-    # for each database
-    for database_name, tasks in databases.items():
+    """
+
+    if "uri" not in entries:
+        msg = f"Missing mandatory 'uri' entry in {database}.{task}.{protocol}.{subset}"
```

Comment:
Did you purposely renamed 'uris' to 'uri' here ?

Reply:
Yes: the point is to make sure keys in `database.yml` and keys in `ProtocolFile`s are consistent.